### ISSUE TITLE
2.3.0 crashes on macOS yosemite

### ISSUE THREAD

**Issue by efx** [keepassxreboot/keepassxc#1671]
## Expected Behavior
Application should open on macOS. I have not tried with 2.3.1.

## Current Behavior
The application crashes immediately.

## Possible Solution
Fix the code signature for libsodium? 

## Steps to Reproduce (for bugs)

1. download https://github.com/keepassxreboot/keepassxc/releases/download/2.3.0/KeePassXC-2.3.0.dmg
2. install application
3. open the application

## Debug Info
```
Process:               KeePassXC [4336]
Path:                  /Applications/KeePassXC.app/Contents/MacOS/KeePassXC
Identifier:            org.keepassx.keepassxc
Version:               ???
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           KeePassXC [4336]
User ID:               502

Date/Time:             2018-03-04 20:16:18.612 -0500
OS Version:            Mac OS X 10.10.5 (14F2511)
Report Version:        11
Anonymous UUID:        C7E1D712-4C65-46C5-68B1-17D3EDDA7754

Sleep/Wake UUID:       14ABA23D-7031-436E-814E-FC6B82806CE9

Time Awake Since Boot: 17000 seconds
Time Since Wake:       560 seconds

Crashed Thread:        0

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: @executable_path/../Frameworks/libsodium.23.dylib
  Referenced from: /Applications/KeePassXC.app/Contents/MacOS/KeePassXC
  Reason: no suitable image found.  Did find:
	/Applications/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib: code signature invalid for '/Applications/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib'

	/Applications/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib: code signature invalid for '/Applications/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib'

Binary Images:
    0x7fff6efbc000 -     0x7fff6eff292f  dyld (353.2.4) <3C068A0D-69D5-3AD6-86D9-6261D3B1564E> /usr/lib/dyld
```

KeePassXC - 2.3.0

Operating system: macOS

**Comment by phoerious** [comment]
Please try 2.3.1.

**Comment by efx** [comment]
This still happens on 2.3.1. Seemingly identical error.

**Comment by rohitksharma** [comment]
I am experiencing similar issues, even after trying 2.3.1. The issue is most common when viewing entries or editing entries within the application. Would it be a good bet to revert back to a previous version until this is resolved?

**Comment by weslly** [comment]
@phoerious https://stackoverflow.com/questions/41865537/how-does-apples-codesign-utility-decide-which-sha-algorithms-to-sign-a-shared

Looks like we need to set `mmacosx-version-min=10.10` on `release-tool`, [though Qt 5.10 still requires at least 10.11](https://doc.qt.io/qt-5.10/supported-platforms.html).

**Comment by peterjw** [comment]
I don't know about Qt 5.10, but KeePassXC 2.3.0 from macports runs fine on Yosemite 10.10. So to my view it's a pure packaging (signing) problem.

**Comment by efx** [comment]
Interesting @peterjw. FWIW I tried installing 2.3.0 using brew and saw the same error.

**Comment by droidmonkey** [comment]
It looks like the libsodium lib is not being signed correctly preventing it from loading

**Comment by weslly** [comment]
@phoerious  I get this if I run `otool -l /Volumes/KeePassXC/KeePassXC.app/Contents/Frameworks/libsodium.23.dylib | grep LC_VERSION_MIN_MACOSX -A2 | uniq`:

```
      cmd LC_VERSION_MIN_MACOSX
  cmdsize 16
  version 10.12
```

But running the same command on the app binary I get `version 10.10` instead, so cmake/clang is only setting the version on the binary and ignoring the libs. We should probably add `-mmacosx-version-min=10.10` to CFLAGS, CXXFLAGS, and LDFLAGS to force the minimum version on the *.dylib files too.

**Comment by efx** [comment]
@weslly sure thing:
```sh
$ codesign -vv /Applications/KeePassXC.app/Contents/Frameworks/libsodium.23.dylib
/Applications/KeePassXC.app/Contents/Frameworks/libsodium.23.dylib: valid on disk
/Applications/KeePassXC.app/Contents/Frameworks/libsodium.23.dylib: satisfies its Designated Requirement
```

**Comment by phoerious** [comment]
If it's just libsodium that's built with the new linker version, then the only thing we can do is build it ourselves. The libsodium.32.dylib is the one from home brew. KeePassXC itself is built with 10.10 compatibility.

**Comment by weslly** [comment]
@phoerious Pretty much all of them are 10.12:

![image](https://user-images.githubusercontent.com/107672/37156037-ada39760-22c3-11e8-95a1-8c54642d923b.png)

**Comment by phoerious** [comment]
Damn. Why do people make backwards compatibility on macOS so damn hard?

**Comment by banerjs** [comment]
I'm trying to get KeePass X installed on a friend's machine and I just ran into this issue. What's the recommended workaround for now? Build from source? If so, could someone post the exact list of steps to follow? I don't work on a Mac myself so I'm quite the novice when it comes to debugging and fixing a build from source issue on that OS.

**Comment by weslly** [comment]
@banerjs https://github.com/keepassxreboot/keepassxc/wiki/Set-up-Build-Environment-on-OS-X and https://github.com/keepassxreboot/keepassxc/wiki/Building-KeePassXC

**Comment by peterjw** [comment]
@banerjs : The .dmg file of one of the old versions - I believe it was 2.2.2- from the Releases page just works on Yosemite or if you happen to have macports installed, just take the macports version.

**Comment by bartatgithub** [comment]
Same problem/error here with fresh install version 2.3.1: 

```
Process:               KeePassXC [23448]
Path:                  /Applications/A/KeePassXC.app/Contents/MacOS/KeePassXC
Identifier:            org.keepassx.keepassxc
Version:               ???
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           KeePassXC [23448]
User ID:               502

Date/Time:             2018-05-08 12:19:31.248 +0200
OS Version:            Mac OS X 10.10.5 (14F2411)
Report Version:        11
Anonymous UUID:        FD463EED-CFC6-8E01-83CD-C88C63CC9306


Time Awake Since Boot: 4700 seconds

Crashed Thread:        0

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: @executable_path/../Frameworks/libsodium.23.dylib
  Referenced from: /Applications/A/KeePassXC.app/Contents/MacOS/KeePassXC
  Reason: no suitable image found.  Did find:
	/Applications/A/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib: code signature invalid for '/Applications/A/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib'

	/Applications/A/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib: code signature invalid for '/Applications/A/KeePassXC.app/Contents/MacOS/../Frameworks/libsodium.23.dylib'
```

Will try the "work arounds" from above.

**Comment by bartatgithub** [comment]
It’s a time ago, so, could you send a link where to find? 
Thanks!

**Comment by peterjw** [comment]
This is still a problem with 2.3.4 released half a year after this ticket was opened. 
If you can't fix the code signature problem for OS X 10.10, I'd recommend not to mark the download as "macOS 10.10 and later"
Just my two pence ...

**Comment by edswangren** [comment]
Same issue here. Not a great user experience...

**Comment by bartatgithub** [comment]
Thanks to work on this problem! 
Meanwhile I have switched to MacOS 10.13.4 and your current version 2.3.4 works smooth. 
Sorry, no running Yosemite anymore to test. 
Maybe someone else?

**Comment by alexblaessle** [comment]
The 2.4.0 snapshot still shows the same behavior.

**Comment by 7usr7local** [comment]
@droidmonkey, thank you for the update.
> Just for grins, can you try the 2.4.0-snapshot build: https://snapshot.keepassxc.org/latest/KeePassXC-2.4.0-snapshot.dmg

Starting it from the DMG image on _10.10.5 Yosemite_ shows the error

![Error Message in german](https://user-images.githubusercontent.com/4259465/51047342-85909b80-15c8-11e9-9c48-9a9cd70f1e7b.png)

(sorry for the german) It basically says rather generic "Can not be opened, contact developer, maybe you need to reinstall the software. Update everything beforehand at all cost. Press _Bericht_ to get more information and send everything to Apple."

Pressing the button indeed gives a form to send information to Apple providinging the lines

```
Process:               KeePassXC [XXXXXX]
Path:                  /Volumes/VOLUME/KeePassXC.app/Contents/MacOS/KeePassXC
Identifier:            org.keepassx.keepassxc
Version:               ???
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           KeePassXC [XXXXX]
User ID:               XXX

Date/Time:             2019-MM-DD HH:MM:SS.068 +XXXX
OS Version:            Mac OS X 10.10.5 (XXXXXX)
Report Version:        11
Crashed Thread:        0
```
[...]
```
Exception** Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Symbol not found: _kSecUseAuthenticationUI
  Referenced from: /Volumes/VOLUME/KeePassXC.app/Contents/MacOS/KeePassXC
  Expected in: /System/Library/Frameworks/Security.framework/Versions/A/Security
 in /Volumes/VOLUME/KeePassXC.app/Contents/MacOS/KeePassXC
```
Followed by some more information about libraries, binary adresses and stuff which I do omit here because I cannot tell wether it contatins sensitive information about my system or not.

Trying to start it from the shell yields basically the same error message.

Not wanting to copy it to `/Applications` for obvious reasons 😏 

Version 2.4 is not (yet) available on MacPorts for Yosemite to crosscheck.

**Comment by droidmonkey** [comment]
Ok great that is very helpful and I know exactly what it wrong.

**Comment by bartatgithub** [comment]
Does this fix also the problem with version 2.4.0 on MacOS 10.13.4? 
MacOS 10.13.4 ---> version 2.3.4 works
MacOS 10.13.4 ---> version 2.4.0 doesn't start up
Thanks!

**Comment by bartatgithub** [comment]
... sorry, version 2.4.1 just installed and tested and works on MacOS 10.13.4! 
Thanks!
